### PR TITLE
159/ Added dark theme in angular application

### PR DESCRIPTION
**Pull request by rajat-vrma**
Issue: [#159](https://github.com/Innostax/tavajs/issues/159)
Added dark mode question in the questioner, 
Fixed modal UI for dark & light mode.
Added custom classes to fix UI issues.
Fixed issue of theme got reset once browser is reloaded

Loom: https://www.loom.com/share/cb1858dc72b241f0bc334f2e718f15df

### CODE REVIEW

File: src/templates/angular/src/index.html [hunk 1]
```diff
@@ -7,7 +7,10 @@
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
 </head>
-<body>
+<body <% if(isDark){%> data-theme="light"<%}%>>
+  <noscript>
+    <p>This page requires JavaScript to work properly. Please enable JavaScript in your browser.</p>
+  </noscript>
```
Comment (manjeetbeniwalinnostax) [inline src/templates/angular/src/index.html:13]: Can we remove this text?

**Review by manjeetbeniwalinnostax — APPROVED** [review]

File: src/index.js [hunk 2]
```diff
@@ -114,6 +114,17 @@ inquirer.prompt(questionnaire).then(async (answers) => {
         }
       );
     }
+    if (isDark && frontEndChoice === 'angular') {
+      fsExtra.copy(
+        `${currentPath}/themeTemplates/angular-themes`,
+        `${frontEnd.path}/src/angular-themes`,
+        (err) => {
+          if (err) {
+            console.log("Error Found:", err);
```
Comment (Chaman-Raghav-Inno) [inline src/index.js:123]: Let's update error messages to be more specific & while keeping them short as well

File: src/templates/angular/src/app/pages/header/header.component.css [hunk 3]
```diff
@@ -0,0 +1,80 @@
+<% if(isDark){%>
+.icon-switcher {
```
Comment (Chaman-Raghav-Inno) [inline src/templates/angular/src/app/pages/header/header.component.css:2]: may be we can use separate files for dark & light modes, & import according to users selection over theme.
like
header.component.light.css
header.component.dark.css

File: src/templates/angular/src/app/pages/header/header.component.html [hunk 4]
```diff
@@ -14,6 +14,19 @@
             <a class="nav-link" [routerLink]="'/users'" routerLinkActive="active">User</a>
           </li>
         </ul>
+        <% if(isDark){%>
+          <form class="d-flex" role="button" [formGroup]="headerForm">
+            <div class="menu">
+              <div class="icon">
+                <label class="icon-switcher" data-js-icon-switcher="data-js-icon-switcher" (change)="handleSelection($event)">
+                  <input id="themeSelector" class="control" type="checkbox" formControlName="themeSelection"/>
```
Comment (Chaman-Raghav-Inno) [inline src/templates/angular/src/app/pages/header/header.component.html:22]: themeSelection
is this holds the value of selected theme?
if yes, lets update its name to selectedTheme /optedTheme
selection is a verb can be use for functions not variables

File: src/templates/angular/src/app/pages/header/header.component.ts [hunk 5]
```diff
@@ -1,15 +1,32 @@
...
+  <% if(isDark){%> isChecked: boolean = true;
+  headerForm!: FormGroup;<%}%>
+  constructor( <% if(isDark){%> private fb: FormBuilder <%}%>) { }
   ngOnInit(): void {
+    <% if(isDark){%>
+    let isThemeSelected: any = sessionStorage.getItem('isDarkModeSelected');
+    let selected = JSON.parse(isThemeSelected);
+    if(selected) $body.setAttribute('data-theme', 'dark')
+    else $body.setAttribute('data-theme', 'light');
+    this.initForm(selected)  <%}%>
```
Comment (Chaman-Raghav-Inno) [inline src/templates/angular/src/app/pages/header/header.component.ts:19]: let's avoid, using a whole chunk of code in conditional statements,
create functions/methods & use them to invoke based on conditional variables.

Comment (Chaman-Raghav-Inno) [reply src/templates/angular/src/app/pages/header/header.component.ts:19]: we can keep this whole chunk of code outside of ngOnInit

File: src/templates/angular/src/app/pages/header/header.component.ts [hunk 6]
```diff
@@ -1,15 +1,32 @@
...
   }
-
+  <% if(isDark){%>
+  initForm(data: any) {
+    this.headerForm = this.fb.group({
+      themeSelection: [data || '']
+    })
+  }
+  handleSelection(event: any) {
+    (event.target.checked) ? $body.setAttribute('data-theme', 'dark') : $body.setAttribute('data-theme', 'light');
```
Comment (Chaman-Raghav-Inno) [inline src/templates/angular/src/app/pages/header/header.component.ts:28]: (event.target.checked) ? $body.setAttribute('data-theme', 'dark') : $body.setAttribute('data-theme', 'light');
=>
$body.setAttribute('data-theme', event.target.checked ? 'dark' : 'light');
lets avoid magic strings

File: src/themeTemplates/angular-themes/dark-theme.css [hunk 7]
```diff
@@ -0,0 +1,31 @@
+
+[data-theme='dark'] {
```
Comment (Chaman-Raghav-Inno) [inline src/themeTemplates/angular-themes/dark-theme.css:2]: what is the significance of this [data-theme='dark'] here?
as this file dedicate for dark theme, do we need to specify this tag over each & every class?

File: src/templates/angular/src/app/pages/header/header.component.css (hunk 3 above)
Comment (rajat-vrma) [reply src/templates/angular/src/app/pages/header/header.component.css:2]: 'isDark' variable contains the Boolean value whether the theme template is selected or not. (not specifically to light or dark mode)

File: src/themeTemplates/angular-themes/dark-theme.css (hunk 7 above)
Comment (rajat-vrma) [reply src/themeTemplates/angular-themes/dark-theme.css:2]: 'data-theme' attribute is added to the body tag, once the user toggles the button to dark mode
T-he value of the data-theme attribute is dynamically updated once clicked on the toggle button for switching between 'light' & 'dark' mode.
- We need to target each & every class once value of the 'data-theme' is updated. 
- As CSS did not support parent-child pattern, or '&', that's why we need to define [data-theme='dark'] selector every time.
Loom: https://www.loom.com/share/847eb2ec9d494e34a531084e35579f32

File: src/templates/angular/src/app/pages/header/header.component.css (hunk 3 above)
Comment (manjeetbeniwalinnostax) [reply src/templates/angular/src/app/pages/header/header.component.css:2]: @Chaman-Raghav-Inno we are updating the **isDark** with **isLightDarkTheme** in upcoming PR

Comment (Chaman-Raghav) [reply src/templates/angular/src/app/pages/header/header.component.css:2]: need changes

File: src/templates/angular/src/app/pages/header/header.component.ts [hunk 8]
```diff
@@ -1,15 +1,41 @@
...
+    <% if(isDark){%>
+    this.selectedTheme()
+    <%}%>
   }
+  <% if(isDark){%>
+  
+  selectedTheme() {
+    let isThemeSelected: any = sessionStorage.getItem('isDarkModeSelected');
+    let selected = JSON.parse(isThemeSelected);
+    if(selected) $body.setAttribute('data-theme', DARK)
```
Comment (Chaman-Raghav) [inline src/templates/angular/src/app/pages/header/header.component.ts:26]: let const, let's update

File: src/themeTemplates/angular-themes/dark-theme.css (hunk 7 above)
Comment (Chaman-Raghav) [reply src/themeTemplates/angular-themes/dark-theme.css:2]: Add TODO: 
for its fix

**Review by Chaman-Raghav — APPROVED** [review]

**Review by Chaman-Raghav — APPROVED** [review]

File: src/themeTemplates/angular-themes/dark-theme.css (hunk 7 above)
Comment (rajat-vrma) [reply src/themeTemplates/angular-themes/dark-theme.css:2]: https://css-tricks.com/almanac/selectors/a/attribute/
image
image

**Pull request merged**